sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/GoneAndRetryWithRetryPolicy.java

```diff
@@ -16,189 +15,324 @@
+        private final static int DEFAULT_WAIT_TIME_IN_SECONDS = 30;
+        private final static int MAXIMUM_BACKOFF_TIME_IN_SECONDS = 15;
+        private final static int INITIAL_BACKOFF_TIME = 1;
+        private final static int BACK_OFF_MULTIPLIER = 2;
+
+        private final RxDocumentServiceRequest request;
+        private volatile int attemptCount = 1;
+        private volatile int attemptCountInvalidPartition = 1;
+        private volatile int currentBackoffSeconds = GoneRetryPolicy.INITIAL_BACKOFF_TIME;
+        private final StopWatch durationTimer;
+        private final int waitTimeInSeconds;
+        //TODO once this is moved to IRetryPolicy, remove from here.
+        public final static Quadruple<Boolean, Boolean, Duration, Integer> INITIAL_ARGUMENT_VALUE_POLICY_ARG = Quadruple.with(false, false,
+            Duration.ofSeconds(60), 0);
+
+        public GoneRetryPolicy(
+            RxDocumentServiceRequest request,
+            Integer waitTimeInSeconds,
+            StopWatch durationTimer) {
+
+            checkNotNull(request, "request must not be null.");
+            this.request = request;
+            this.durationTimer = durationTimer;
+            this.waitTimeInSeconds = waitTimeInSeconds != null ? waitTimeInSeconds : DEFAULT_WAIT_TIME_IN_SECONDS;
+        }
+
+        private boolean isRetryableException(Exception exception) {
+            if (exception instanceof GoneException ||
+                exception instanceof RetryWithException ||
+                exception instanceof PartitionIsMigratingException ||
+                exception instanceof PartitionKeyRangeIsSplittingException) {
+
+                return true;
+            }
+
+            if (exception instanceof InvalidPartitionException) {
+                return this.request.getPartitionKeyRangeIdentity() == null ||
+                    this.request.getPartitionKeyRangeIdentity().getCollectionRid() == null;
+            }
+
+            return false;
+        }
+
+        private CosmosException logAndWrapExceptionWithLastRetryWithException(Exception exception) {
+            RetryWithException lastRetryWithException = this.request.getLastRetryWithException();
+
+            String exceptionType;
+            if (exception instanceof GoneException) {
+                exceptionType = "GoneException";
+            } else if (exception instanceof PartitionKeyRangeGoneException) {
+                exceptionType = "PartitionKeyRangeGoneException";
+            } else if (exception instanceof  InvalidPartitionException) {
+                exceptionType = "InvalidPartitionException";
+            } else if (exception instanceof  PartitionKeyRangeIsSplittingException) {
+                exceptionType = "PartitionKeyRangeIsSplittingException";
+            } else if (exception instanceof CosmosException) {
+                logger.warn("Received CosmosException after backoff/retry. Will fail the request.",
+                    exception);
```

Comment:
what happen if the code throws unrelated exception to this retry policy? do we log it here?

if so an exception will get logged here and elsewhere where it is truely handled. (e.g., 429). We shouldn't log unrelated exceptions to this policy and let related other policy to handle or log the exception.

Reply:
Same behavior as before - should we really change it - what is the harm?

Reply:
thanks for clarification. I thought this is new. LGTM.